src/d3a/models/strategy/external_strategies/storage.py

```diff
@@ -0,0 +1,261 @@
+        except Exception as e:
+            logging.error(f"Error when handling list bids on area {self.device.name}: "
+                          f"Exception: {str(e)}")
+            self.redis.publish_json(
+                list_bids_response_channel,
+                {"status": "error",
+                 "error_message": f"Error when listing bids on area {self.device.name}."})
+
+    def _delete_bid(self, payload):
+        delete_bid_response_channel = f'{self.device.name}/delete_bid/response'
+        if not check_for_connected_and_reply(self.redis,
+                                             delete_bid_response_channel, self.connected):
+            return
+        try:
+            arguments = json.loads(payload["data"])
+            assert set(arguments.keys()) == {'bid'}
+        except Exception:
+            self.redis.publish_json(
+                delete_bid_response_channel,
+                {"error": "Incorrect delete bid request. Available parameters: (bid)."}
+            )
+        else:
+            self.pending_requests.append(
+                IncomingRequest("delete_bid", arguments, delete_bid_response_channel))
+
+    def _delete_bid_impl(self, arguments, response_channel):
+        try:
+            self.remove_bid_from_pending(arguments["bid"], self.market.id)
+            self.redis.publish_json(response_channel,
+                                    {"status": "ready", "bid_deleted": arguments["bid"]})
+        except Exception as e:
+            logging.error(f"Error when handling bid delete on area {self.device.name}: "
+                          f"Exception: {str(e)}, Bid Arguments: {arguments}")
+            self.redis.publish_json(
+                response_channel,
+                {"status": "error",
+                 "error_message": f"Error when handling bid delete "
+                                  f"on area {self.device.name} with arguments {arguments}."})
+
+    def _bid(self, payload):
+        bid_response_channel = f'{self.device.name}/bid/response'
+        if not check_for_connected_and_reply(self.redis, bid_response_channel, self.connected):
+            return
+        try:
+            arguments = json.loads(payload["data"])
+            assert set(arguments.keys()) == {'price', 'energy'}
+            arguments['buyer_origin'] = self.device.name
+        except Exception:
+            self.redis.publish_json(
+                bid_response_channel,
+                {"error": "Incorrect bid request. Available parameters: (price, energy)."}
+            )
+        else:
+            self.pending_requests.append(
+                IncomingRequest("bid", arguments, bid_response_channel))
+
+    def _bid_impl(self, arguments, bid_response_channel):
+        try:
+            self.state.clamp_energy_to_buy_kWh([self.market.time_slot])
+            max_energy = self.state.energy_to_buy_dict[self.market.time_slot]
+            assert arguments["energy"] <= max_energy
```

Comment:
Please do not assert here, and move this check in the _bid method, and also validate that the error response message is sent to the user. The general pattern here is that, validitation of the api arguments should be performed in the _bid, _offer .... methods (and respond with an error to the user), and handle errors when performing market operations in the *_impl methods.

Reply:
To my eyes this assert protects the [bid](https://github.com/gridsingularity/d3a/pull/639/files#diff-1d553c99bd95475906818e1177a93f29R201) to be posted in first place. Following the other approach would also need to delete the posted bid too.

Reply:
I do not want to delete the bid actually. I would just like to move this check on the _bid method, instead of the _bid_impl method, in order to not even try to post an offer if the assert is violated.